src/simsopt/mhd/vmec_diagnostics.py

```diff
@@ -986,18 +994,147 @@ def residual(theta_v, phi0, theta_p_target, jradius):
+            a[inds] = b[inds]
+            b[inds] = c[inds]
+            c[inds] = a[inds]
+            fa[inds] = fb[inds]
+            fb[inds] = fc[inds]
+            fc[inds] = fa[inds]
+            
+            tol1 = 2.0 * eps * np.abs(b) + 0.5 * tol
+            Xm = 0.5 * (c - b)
+            indsC = ((np.abs(Xm) <= tol1) + (fb == 0.0))
+    
+            roots[indsC] = b[indsC]
+            convds[indsC] = True
+            nconv = np.sum(indsC)
+            if nconv == toteval:
+                converged = True
+                return roots, converged, nconv
+            
+            p = 0 * a
+            s = 0 * p
+            q = 0 * p
+            r = 0 * p
+    
+            inds1 = (((np.abs(e) >= tol1) * (np.abs(fa) > np.abs(fb))))
+            s[inds1] = (fb[inds1])/(fa[inds1])
+    
+            inds2 = (a == c)
+            inds = inds2*inds1
+            s[inds] = (fb[inds])/(fa[inds])
+            p[inds] = 2.0 * Xm[inds] * s[inds]
+            q[inds] = 1.0 - s[inds]
+    
+            inds = np.invert(inds2)*inds1
+            q[inds] = fa[inds] / fc[inds]
+            r[inds] = fb[inds] / fc[inds]
+            p[inds] = s[inds] * (2.0 * Xm[inds] * q[inds] * (q[inds] - r[inds]) - (b[inds] - a[inds]) * (r[inds] - 1.0))
+            q[inds] = (q[inds] - 1.0) * (r[inds] - 1.0) * (s[inds] - 1.0)
+    
+            inds = (p > 0.0)*inds1
+            q[inds] = -q[inds]
+            p[inds1] = np.abs(p[inds1])
+    
+            inds2 = (2.0 * p < np.minimum(3.0 * Xm * q - np.abs(tol1 * q), np.abs(e * q)))
+            inds = inds2*inds1
+            e[inds] = d[inds]
+            d[inds] = (p[inds])/(q[inds])
+    
+            inds = np.invert(inds2)*inds1
+            d[inds] = Xm[inds]
+            e[inds] = d[inds]
+            
+            inds = np.invert(inds1)
+            d[inds] = Xm[inds]
+            e[inds] = d[inds]
+    
+            a = 1 * b
+            fa = 1 * fb
+    
+            inds = np.where((np.abs(d) > tol1))
+            b[inds] += d[inds]
+    
+            inds = np.where((np.abs(d) <= tol1))
+            b[inds] += np.copysign(tol1[inds], Xm[inds])
+            fb = fzero_residuals_function(b, theta_pest_targets, phis, xm, xn, lmns, lmnc, lasym)
+        return roots, converged, nconv
+    
+    #theta_vmec = np.zeros((ns, nalpha, nl))
+    #for js in range(ns):
+    #    for jalpha in range(nalpha):
+    #        for jl in range(nl):
+    #            theta_guess = theta_pest[js, jalpha, jl]
+    #            solution = root_scalar(residual,args=(phi[js, jalpha, jl], theta_pest[js, jalpha, jl], js),
+    #                                   bracket=(theta_guess - 1.0, theta_guess + 1.0))
+    #            theta_vmec[js, jalpha, jl] = solution.root
+    ## Solve for theta_vmec corresponding to theta_pest (new routine)
+    ## Does the same calculation as the commented code above but faster
     theta_vmec = np.zeros((ns, nalpha, nl))
     for js in range(ns):
```

Comment:
If you can get rid of the loop over l, can you get rid of the loops over s and alpha too?

Reply:
Yes, but the overall speedup will not be significant. For example, speeding up vectorized Newton by a factor of two would correspond to < 10% overall speedup.
```
Time taken to specify all arrays = 0.0372 s 
Time taken by vectorized Newton = 0.0467 s
Time taken by Fourier sum = 0.1972 s
```
Maybe for a future version.